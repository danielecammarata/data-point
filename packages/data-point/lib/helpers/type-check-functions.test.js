/* eslint-env jest */

const {
  isString,
  isNumber,
  isBoolean,
  isFunction,
  isError,
  isArray,
  isObject
} = require("./type-check-functions").defaults;

const { createTypeCheckReducer } = require("./type-check-functions");

function testTypeChecker(typeCheck, goodValue, badValue) {
  test("It should return true if type match", () => {
    expect(typeCheck(goodValue)).toBe(goodValue);
  });
  test(`It should throw error when not matched`, () => {
    expect(() => {
      typeCheck(badValue);
    }).toThrowErrorMatchingSnapshot();
  });
}

describe("isString", () => {
  testTypeChecker(isString, "foo", 1);
});

describe("isNumber", () => {
  testTypeChecker(isNumber, 1, "foo");
});

describe("isBoolean", () => {
  testTypeChecker(isBoolean, true, "foo");
  testTypeChecker(isBoolean, false, "foo");
});

describe("isFunction", () => {
  const fn = () => {};
  testTypeChecker(isFunction, fn, "foo");
});

describe("isError", () => {
  testTypeChecker(isError, new Error(), "foo");
});

describe("isArray", () => {
  testTypeChecker(isArray, [], "foo");
});

describe("isObject", () => {
  testTypeChecker(isObject, {}, []);
});

describe("custom type check reducer", () => {
  const expectedType = "non-empty array";

  const check = input => {
    return Array.isArray(input) && !!input.length;
  };

  const checkWithCustomMessage = input => {
    return check(input) || "Hello the world is collapsing.";
  };

  it("should return the input when it's valid", () => {
    const reducer = createTypeCheckReducer(check, expectedType);
    expect(reducer([true])).toEqual([true]);
  });

  it("should throw error for invalid input", () => {
    const reducer = createTypeCheckReducer(check);
    expect(() => reducer(false)).toThrowErrorMatchingSnapshot();
  });

  it("should throw error (with expected type)", () => {
    const reducer = createTypeCheckReducer(check, expectedType);
    expect(() => reducer(false)).toThrowErrorMatchingSnapshot();
  });

  it("should throw error (with custom message)", () => {
    const reducer = createTypeCheckReducer(checkWithCustomMessage);
    expect(() => reducer(false)).toThrowErrorMatchingSnapshot();
  });
});
