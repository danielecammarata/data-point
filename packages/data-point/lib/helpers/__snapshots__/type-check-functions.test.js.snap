// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom type check reducer should throw error (with custom message) 1`] = `
"Entity type check failed!
Actual type: boolean
Input value: false
Hello the world is collapsing."
`;

exports[`custom type check reducer should throw error (with expected type) 1`] = `
"Entity type check failed!
Expected type: non-empty array
Actual type: boolean
Input value: false"
`;

exports[`custom type check reducer should throw error for invalid input 1`] = `
"Entity type check failed!
Actual type: boolean
Input value: false"
`;

exports[`isArray It should throw error when not matched 1`] = `
"Entity type check failed!
Expected type: array
Actual type: string
Input value: 'foo'"
`;

exports[`isBoolean It should throw error when not matched 1`] = `
"Entity type check failed!
Expected type: boolean
Actual type: string
Input value: 'foo'"
`;

exports[`isBoolean It should throw error when not matched 2`] = `
"Entity type check failed!
Expected type: boolean
Actual type: string
Input value: 'foo'"
`;

exports[`isError It should throw error when not matched 1`] = `
"Entity type check failed!
Expected type: error
Actual type: string
Input value: 'foo'"
`;

exports[`isFunction It should throw error when not matched 1`] = `
"Entity type check failed!
Expected type: function
Actual type: string
Input value: 'foo'"
`;

exports[`isNumber It should throw error when not matched 1`] = `
"Entity type check failed!
Expected type: number
Actual type: string
Input value: 'foo'"
`;

exports[`isObject It should throw error when not matched 1`] = `
"Entity type check failed!
Expected type: object
Actual type: array
Input value: []"
`;

exports[`isString It should throw error when not matched 1`] = `
"Entity type check failed!
Expected type: string
Actual type: number
Input value: 1"
`;
